blog/content/post/2025-07-xx-edk2-on-odroid-m2.md

```diff
@@ -0,0 +1,325 @@
+
+```c
+static struct regulator_init_data rk806_init_data[] = {
+    /* Master PMIC */
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK1, 950000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK2, 950000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK3, 750000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK4, 950000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK5, 900000),
+    /* This is not configured in the M2's Linux device tree
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK6, 1100000), */
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK7, 2000000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK8, 3300000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK10, 1800000),
+
+    RK8XX_VOLTAGE_INIT(MASTER_NLDO1, 750000),
+    RK8XX_VOLTAGE_INIT(MASTER_NLDO2, 900000),
+    RK8XX_VOLTAGE_INIT(MASTER_NLDO3, 837500),
+    RK8XX_VOLTAGE_INIT(MASTER_NLDO4, 850000),
+    /* RK8XX_VOLTAGE_INIT(MASTER_NLDO5, 750000),*/
+
+PcieIoInit (
+  UINT32  Segment
+  )
+{
+  /* Set reset to gpio output mode */
+  if (Segment == PCIE_SEGMENT_PCIE20L2) {
+    // M.2 M Key
+    GpioPinSetDirection (3, GPIO_PIN_PD1, GPIO_PIN_OUTPUT);
+  }
+}
+
+VOID
+EFIAPI
+PciePowerEn (
+  UINT32   Segment,
+  BOOLEAN  Enable
+  )
+{
+  /* nothing to power on */
+}
+
+VOID
+EFIAPI
+PciePeReset (
+  UINT32   Segment,
+  BOOLEAN  Enable
+  )
+{
+  if (Segment == PCIE_SEGMENT_PCIE20L2) {
+    GpioPinWrite (3, GPIO_PIN_PD1, !Enable);
+  }
+}
+```
+
+From this we can gather a couple of things:
+
+- There is a M.2 M-key slot on the board
+- There is no Power Enable signal for PCIe on this board
+- There is a PCIe Reset signal, in this case connected to GPIO_PIN_PD1 on GPIO
+  controller 3
+
+We need to understand how all this relates to the Odroid M2 we're working with.
+Thankfully, Hardkernel provides us with schematics for this board.
+
+[ insert schematic here ]
+
+We can see that on this board:
+
+- There is an M.2 slot on the board
+- There is a power enable signal on GPIO0_PC6
+- There is a PCIe reset signal GPIO1_PA7
+
+So the code for our board should look like this:
+
+```c
+    EFIAPI
+    PcieIoInit(
+        UINT32 Segment)
+{
+  /* Set power enable and reset to gpio output mode */
+  if (Segment == PCIE_SEGMENT_PCIE20L2)
+  {
+    // M.2 M Key
+    GpioPinSetDirection(1, GPIO_PIN_PA7, GPIO_PIN_OUTPUT);
+    GpioPinSetDirection(0, GPIO_PIN_PC6, GPIO_PIN_OUTPUT);
+  }
+}
+
+VOID
+    EFIAPI
+    PciePowerEn(
+        UINT32 Segment,
+        BOOLEAN Enable)
+{
+  if (Segment == PCIE_SEGMENT_PCIE20L2)
+  {
+    GpioPinWrite(0, GPIO_PIN_PC6, Enable);
+  }
+}
+
+VOID
+    EFIAPI
+    PciePeReset(
+        UINT32 Segment,
+        BOOLEAN Enable)
+{
+  if (Segment == PCIE_SEGMENT_PCIE20L2)
+  {
+    GpioPinWrite(1, GPIO_PIN_PA7, !Enable);
+  }
+}
+```
+
+### Fan control
+
+The Odroid has a small fan installed on the CPU heatsink that helps remove
+excess heat from the SoC. We should configure it properly to ensure the
+processor doesn't overheat.
+
+The PWM signal that controls fan speed is connected to pin GPIO1_PA2, which
```

Comment:
We reading this from schematics?

Reply:
Yes, added a screenshot https://github.com/3mdeb/news-and-ideas/pull/674/commits/f5f14ad5f29048f5b647c8c555fafaf0ae6f4f74